File: backend/src/main/java/com/autocare/service/VehicleService.java
package com.autocare.service;

import com.autocare.dto.VehicleRequest;
import com.autocare.dto.VehicleResponse;
import com.autocare.entity.Reminder;
import com.autocare.entity.Role;
import com.autocare.entity.User;
import com.autocare.entity.Vehicle;
import com.autocare.exception.ConflictException;
import com.autocare.exception.ForbiddenException;
import com.autocare.exception.ResourceNotFoundException;
import com.autocare.repository.ReminderRepository;
import com.autocare.repository.UserRepository;
import com.autocare.repository.VehicleRepository;
import com.autocare.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class VehicleService {

    private final VehicleRepository vehicleRepository;
    private final UserRepository userRepository;
    private final ReminderRepository reminderRepository;

    @Transactional
    public VehicleResponse registerVehicle(VehicleRequest request, UserPrincipal principal) {
        if (vehicleRepository.existsByVin(request.getVin())) {
            throw new ConflictException("A vehicle with VIN " + request.getVin() + " is already registered.");
        }

        User owner = userRepository.findById(principal.getUserId())
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        Vehicle vehicle = Vehicle.builder()
            .owner(owner)
            .vin(request.getVin().trim().toUpperCase())
            .make(request.getMake().trim())
            .model(request.getModel().trim())
            .year(request.getYear())
            .odometer(request.getOdometer())
            .build();

        Vehicle savedVehicle = vehicleRepository.save(vehicle);

        // Provision initial predictive maintenance reminder for new vehicle
        Reminder initialReminder = Reminder.builder()
            .vehicle(savedVehicle)
            .reminderType("Routine Inspection & Diagnostics")
            .dueDate(LocalDate.now().plusMonths(3))
            .message("Quarterly multi-point vehicle inspection and fluid level check.")
            .status("ACTIVE")
            .build();
        reminderRepository.save(initialReminder);

        return mapToResponse(savedVehicle);
    }

    @Transactional(readOnly = true)
    public List<VehicleResponse> getVehicles(UserPrincipal principal) {
        if (principal.getRole() == Role.CUSTOMER) {
            return vehicleRepository.findAllByOwner_UserId(principal.getUserId()).stream()
                .map(this::mapToResponse)
                .toList();
        } else {
            return vehicleRepository.findAllByWorkshopId(principal.getWorkshopId()).stream()
                .map(this::mapToResponse)
                .toList();
        }
    }

    @Transactional(readOnly = true)
    public VehicleResponse getVehicleById(Integer vehicleId, UserPrincipal principal) {
        Vehicle vehicle = vehicleRepository.findById(vehicleId)
            .orElseThrow(() -> new ResourceNotFoundException("Vehicle not found with ID: " + vehicleId));

        // Enforce tenant boundary
        if (!vehicle.getOwner().getWorkshop().getWorkshopId().equals(principal.getWorkshopId())) {
            throw new ForbiddenException("Unauthorized access to vehicle outside your workshop tenant.");
        }

        if (principal.getRole() == Role.CUSTOMER && !vehicle.getOwner().getUserId().equals(principal.getUserId())) {
            throw new ForbiddenException("Unauthorized access to vehicle owned by another user.");
        }

        return mapToResponse(vehicle);
    }

    private VehicleResponse mapToResponse(Vehicle v) {
        return VehicleResponse.builder()
            .vehicleId(v.getVehicleId())
            .ownerId(v.getOwner().getUserId())
            .ownerName(v.getOwner().getFirstName() + " " + v.getOwner().getLastName())
            .vin(v.getVin())
            .make(v.getMake())
            .model(v.getModel())
            .year(v.getYear())
            .odometer(v.getOdometer())
            .createdAt(v.getCreatedAt())
            .build();
    }
}

File: backend/src/services/vehicleService.js
/**
 * AutoCare AI - Vehicle Service
 * Handles vehicle registration, ownership isolation, and predictive maintenance provisioning
 */

const { withTransaction } = require('../config/db');
const vehicleRepo = require('../repositories/vehicleRepo');
const reminderRepo = require('../repositories/reminderRepo');
const {
  BadRequestError,
  NotFoundError,
  ForbiddenError,
  ConflictError
} = require('../middleware/errorHandler');

class VehicleService {
  /**
   * Helper: Add months to current date formatted as YYYY-MM-DD
   */
  getFutureDate(months) {
    const d = new Date();
    d.setMonth(d.getMonth() + months);
    return d.toISOString().split('T')[0];
  }

  /**
   * Register a new vehicle for the authenticated customer.
   */
  async registerVehicle({ vin, make, model, year, odometer }, userPrincipal) {
    if (!vin || !make || !model || !year || odometer === undefined) {
      throw new BadRequestError('All vehicle specifications (VIN, Make, Model, Year, Odometer) are required.');
    }

    const normalizedVin = vin.trim().toUpperCase();
    if (await vehicleRepo.existsByVin(null, normalizedVin)) {
      throw new ConflictError(`A vehicle with VIN ${normalizedVin} is already registered.`);
    }

    return withTransaction(async (conn) => {
      // 1. Create vehicle row
      const vehicleId = await vehicleRepo.create(conn, {
        ownerId: userPrincipal.userId,
        vin: normalizedVin,
        make: make.trim(),
        model: model.trim(),
        year: parseInt(year, 10),
        odometer: parseInt(odometer, 10)
      });

      // 2. Provision initial quarterly maintenance reminder
      const dueDate = this.getFutureDate(3);
      await reminderRepo.create(conn, {
        vehicleId,
        reminderType: 'Routine Inspection & Diagnostics',
        dueDate,
        message: 'Quarterly multi-point vehicle inspection and fluid level check.',
        status: 'ACTIVE'
      });

      // 3. Fetch newly created vehicle with joined attributes
      const vehicle = await vehicleRepo.findById(conn, vehicleId);
      return {
        vehicleId: vehicle.vehicleId,
        ownerId: vehicle.ownerId,
        ownerName: vehicle.ownerName,
        vin: vehicle.vin,
        make: vehicle.make,
        model: vehicle.model,
        year: vehicle.year,
        odometer: vehicle.odometer,
        createdAt: vehicle.createdAt
      };
    });
  }

  /**
   * Fetch vehicles accessible by the authenticated user.
   */
  async getVehicles(userPrincipal) {
    if (userPrincipal.role === 'CUSTOMER') {
      return vehicleRepo.findAllByOwnerId(null, userPrincipal.userId);
    } else {
      return vehicleRepo.findAllByWorkshopId(null, userPrincipal.workshopId);
    }
  }

  /**
   * Fetch single vehicle by ID with tenant & ownership boundary checks.
   */
  async getVehicleById(vehicleId, userPrincipal) {
    const vehicle = await vehicleRepo.findById(null, vehicleId);
    if (!vehicle) {
      throw new NotFoundError(`Vehicle not found with ID: ${vehicleId}`);
    }

    if (vehicle.workshopId !== userPrincipal.workshopId) {
      throw new ForbiddenError('Unauthorized access to vehicle outside your workshop tenant.');
    }

    if (userPrincipal.role === 'CUSTOMER' && vehicle.ownerId !== userPrincipal.userId) {
      throw new ForbiddenError('Unauthorized access to vehicle owned by another user.');
    }

    return {
      vehicleId: vehicle.vehicleId,
      ownerId: vehicle.ownerId,
      ownerName: vehicle.ownerName,
      vin: vehicle.vin,
      make: vehicle.make,
      model: vehicle.model,
      year: vehicle.year,
      odometer: vehicle.odometer,
      createdAt: vehicle.createdAt
    };
  }
}

module.exports = new VehicleService();
